Дописал переводчик

alozhka opened the pull request.

alozhka committed:
Дописал переводчик

pklgn commented on Dictionary/Program.cs:
```diff
@@ -0,0 +1,188 @@
+    private static void Translate()
+    {
+        Console.Write("Введите слово: ");
+        var word = Console.ReadLine();
```
Не используем `var`.

pklgn commented on Dictionary/Program.cs:
```diff
@@ -0,0 +1,188 @@
+    public static void Main(string[] args)
+    {
+        ShowMenu();
+        var command = Console.ReadLine();
```
Не используем `var`.

pklgn commented on Dictionary/Program.cs:
```diff
@@ -0,0 +1,188 @@
+    {
+        Console.Write("Введите слово: ");
+        var word = Console.ReadLine();
+        if (word is null || !Regex.IsMatch(word, "^[а-яёА-Я]+$"))
```
Если добавили малую букву ё, то почему бы не добавить большую Ё?

pklgn commented on Dictionary/Program.cs:
```diff
@@ -0,0 +1,188 @@
+    private static void RemoveTranslation()
+    {
+        Console.Write("Введите слово: ");
+        var word = Console.ReadLine();
```
Не используем `var`.

pklgn commented on Dictionary/Program.cs:
```diff
@@ -0,0 +1,188 @@
+    {
+        Console.Write("Введите слово: ");
+        var word = Console.ReadLine();
+        if (word is null || !Regex.IsMatch(word, "^[а-яёА-Я]+$"))
```
Может вынести строку `^[а-яёА-Я]+$` в константу?

pklgn commented on Dictionary/Program.cs:
```diff
@@ -0,0 +1,188 @@
+    {
+        Console.Write("Введите слово: ");
+        var word = Console.ReadLine();
+        if (word is null || !Regex.IsMatch(word, "^[а-яёА-Я]+$"))
```
Общая логика, встречается в разных методах, стоит вынести в отдельный метод и переиспользовать.

pklgn commented on Dictionary/Program.cs:
```diff
@@ -0,0 +1,188 @@
+        Console.WriteLine("Успешно.\n");
+    }
+
+    private static MyDictionary LoadData()
```
Путь к файлу стоит сделать передаваемым через параметры. Сам файл можно вынести в константу и передавать ее при вызове метода.

pklgn commented on Dictionary/Program.cs:
```diff
@@ -0,0 +1,188 @@
+            var data = JsonSerializer.Deserialize<MyDictionary>(file);
+            return data ?? [];
+        }
+        catch (Exception)
```
Почему тут обрабатываем базовое исключение, а не конкретное вроде `JsonException`?

pklgn commented on Dictionary/Program.cs:
```diff
@@ -0,0 +1,188 @@
+
+    private static MyDictionary LoadData()
+    {
+        using var file = new FileStream("translations.json", FileMode.OpenOrCreate, FileAccess.Read);
```
Здесь и ниже не используем `var`)

pklgn commented on Dictionary/Program.cs:
```diff
@@ -0,0 +1,188 @@
+            return [];
+        }
+    }
+    private static void SaveData()
```
Путь к файлу стоит передавать через параметры. Сейчас, если нужно изменить название сохраняемого и читаемого файла, то придется править в двух местах, есть риск забыть исправить в одном из методов, либо опечататься при переименовании в одном из методов. С введением константы и передачей ее значения при вызове `SaveData` и `LoadData` такие риски отсутствуют: правим константу и не трогаем сами методы.

alozhka committed:
Исправил ошибки с ревью

Сделал общий метод под парсинг, вынес регулярные выражения в константы, сделал выбор файла через параметры

pklgn commented on Dictionary/Program.cs:
```diff
@@ -0,0 +1,152 @@
+    {
+        string word = ParsingHelper.GetRussianWordFromConsole();
+
+        if (!_translations.TryGetValue(word, out var value))
```
Вместо `value` по аналогии с `ChangeTranslation` напрашивается идентификатор `translation`.

pklgn commented on Dictionary/ParsingHelper.cs:
```diff
@@ -0,0 +1,44 @@
+    {
+        while (true)
+        {
+            var word = Console.ReadLine();
```
Здесь и далее не используем `var`. Давай возьмем это за правило. В любых проектах не использовать `var`.
Можно попробовать настроить проверку и добавить в проект, например, stylecop для того, чтобы анализатор подсказывал сразу подобные исправления. Но в целом проекты небольшие, думаю, это будет излишним. Пока попробуем обойтись без этого.

pklgn commented on Dictionary/ParsingHelper.cs:
```diff
@@ -0,0 +1,44 @@
+        while (true)
+        {
+            var word = Console.ReadLine();
+            if (word is null || !Regex.IsMatch(word, regex))
```
А если слово введено будет с большими буквами? `ToLower()` вызывается уже после проверки. Может стоит его вызвать сразу после `ReadLine` в таком случае?

pklgn commented:
Осталось несколько доработок. Отдельно зафиксирую, что `var` не используем: ни в этом проекте, ни в любых других)
Еще предлагаю отписывать в ответах на комментарии ревью свои мысли. Может ты не согласен с каким-то доводом и хочешь его обсудить ("На мой взгляд, здесь будет лучше...") или согласен и ты поправил код в соответствии с комментарием (можно поставить лайк на коммент). Мне кажется, так легче отслеживать, что осталось поправить, и мне понятнее будет ход мыслей.

alozhka commented:
Хорошо. Про var я понял, просто тут забыл убрать один. 
Я согласен с тем, что тяжело определить тип, который получаешь например от метода. То есть мы, смотря на название метода, не можем с виду понять, какой тип возвращается, и поэтому при создании переменной явно указываем тип. 
Но если мы создаем новый объект, то по сути тип переменной является названием класса. Если вдруг у меня будет очень длинное название типа, то думаю в этом варианте вполне логично будет использовать var.
Например в этом случае:

var dict = new Dictionary<string, IReadOnlyList<string>>();

Думаю, тут var в самый раз. Могу ли я в этих случаях его использовать?

pklgn commented:
> var dict = new Dictionary<string, IReadOnlyList>();
> 
> Думаю, тут var в самый раз. Могу ли я в этих случаях его использовать?

В таких случаях поступают наоборот)
Пишут 
`Dictionary<string, IReadOnlyList> dict = new();`
В IDE это называется вроде `use target-type`

![image_2024-04-06_23-14-42](https://github.com/alozhka/travelline_practice/assets/73781985/348c4a08-7713-4d28-a975-2a8ac8bcbf6a)

alozhka committed:
Избавился от var и исправил проблему с верхним регистром

pklgn reviewed (approved):
LGTM!
